src/vs/base/browser/performance.ts

```diff
@@ -0,0 +1,231 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+export namespace inputLatency {
+
+	// Measurements are recorded in typed arrays with a fixed buffer length to their own impact on
+	// input latency.
+	const enum Constants {
+		bufferLength = 256
+	}
+	const measurementsKeydown = new Float32Array(Constants.bufferLength);
+	const measurementsInput = new Float32Array(Constants.bufferLength);
+	const measurementsRender = new Float32Array(Constants.bufferLength);
+	const measurementsInputLatency = new Float32Array(Constants.bufferLength);
+	let measurementsCount = 0;
+
+	// The state of each event, this helps ensure the integrity of the measurement and that
+	// something unexpected didn't happen that could skew the measurement.
+	const enum EventPhase {
+		Before = 0,
+		InProgress = 1,
+		Finished = 2
+	}
+	const state = {
+		keydown: EventPhase.Before,
+		input: EventPhase.Before,
+		render: EventPhase.Before,
+		selection: EventPhase.Before
+	};
+
+	/**
+	 * Mark the start of the keydown event.
+	 */
+	export function markKeydownStart() {
+		performance.mark('inputlatency/start');
+		performance.mark('keydown/start');
+		state.keydown = EventPhase.InProgress;
+		queueMicrotask(() => markKeydownEnd());
+	}
+
+	/**
+	 * Mark the end of the keydown event.
+	 */
+	function markKeydownEnd() {
+		// Only measure the first render after keyboard input
+		performance.mark('keydown/end');
+		state.keydown = EventPhase.Finished;
+	}
+
+	/**
+	 * Mark the start of the input event.
+	 */
+	export function markInputStart() {
+		performance.mark('input/start');
+		state.input = EventPhase.InProgress;
+	}
+
+	/**
+	 * Mark the end of the input event.
+	 */
+	export function markInputEnd() {
+		queueMicrotask(() => {
+			performance.mark('input/end');
+			state.input = EventPhase.Finished;
+		});
+	}
+
+	/**
+	 * Mark the start of the animation frame performing the rendering.
+	 */
+	export function markRenderStart() {
+		// Render may be triggered during input, but we only measure the following animation frame
+		if (state.keydown === EventPhase.Finished && state.input === EventPhase.Finished && state.render === EventPhase.Before) {
+			// Only measure the first render after keyboard input
+			performance.mark('render/start');
+			state.render = EventPhase.InProgress;
+			queueMicrotask(() => markRenderEnd());
+		}
+	}
+
+	/**
+	 * Mark the end of the animation frame performing the rendering.
+	 *
+	 * An input latency sample is complete when both the textarea selection change event and the
+	 * animation frame performing the rendering has triggered.
+	 */
```

Comment:
What is the purpose of the textarea selection change event? I think we keep the textarea empty on Windows & Linux when using `"editor.accessibilitySupport": "off"`. IIRC the reason we listen to the selection change event is to capture the case when users are using Braille devices to move the cursor inside the textarea (but in the case when the textarea has content). My question is:
* Why is the time when the browser emits a selection change event in the `<textarea>` relevant here?
* What should we do if this event is not reliable? e.g. the browser doesn't fire it because the selection remains unchanged at a certain offset.

Reply:
I added that because to work around some edge case. It was a while ago, but I think it was to do with 2 render events occurring, but the second one was the one that actually did it. So not having the selection check there causes the latency to appear too low.

> What should we do if this event is not reliable?

Do you have a browser in mind? It seemed to reliably fire every time I was testing

Reply:
My concern is that this inflates all latency numbers. From the diagram you posted below it looks like `inputlatency` now contains a lot of other stuff after rendering has finished (i.e. it is increased by around 20%):

![image](https://user-images.githubusercontent.com/5047891/201381182-d02c6fc8-b57a-4fb1-8246-ca962ee75c31.png)

Reply:
It may have been before I did the `setImmediate` thing since microtasks will run before layout/paint/composite. Maybe if we just listen for the first render (haven't seen that double render thing in a while so could have been something else) and use `setImmediate` then that's the best of both worlds? We could ship that and check what numbers come in?